--- a/drivers/net/wireless/ath/ath9k/common-init.c
+++ b/drivers/net/wireless/ath/ath9k/common-init.c
@@ -86,6 +86,17 @@ static const struct ieee80211_channel at
 	CHAN5G(5785, 35), /* Channel 157 */
 	CHAN5G(5805, 36), /* Channel 161 */
 	CHAN5G(5825, 37), /* Channel 165 */
+	/* ITS-G5 channels */
+	/* ITS-G5B */
+	CHAN5G(5860, 38), /* Channel 172 - G5-SCH4 */
+	CHAN5G(5870, 39), /* Channel 174 - G5-SCH3 */
+	/* ITS-G5A */
+	CHAN5G(5880, 40), /* Channel 176 - G5-SCH1 */
+	CHAN5G(5890, 41), /* Channel 178 - G5-SCH2 */
+	CHAN5G(5900, 42), /* Channel 180 - G5-CCH  */
+	/* ITS-G5D */
+	CHAN5G(5910, 43), /* Channel 182 - G5-SCH5 */
+	CHAN5G(5920, 44), /* Channel 184 - G5-SCH6 */
 };
 
 /* Atheros hardware rate code addition for short preamble */
--- a/drivers/net/wireless/ath/ath9k/hw.h
+++ b/drivers/net/wireless/ath/ath9k/hw.h
@@ -74,7 +74,7 @@
 
 #define ATH9K_RSSI_BAD			-128
 
-#define ATH9K_NUM_CHANNELS	38
+#define ATH9K_NUM_CHANNELS	45
 
 /* Register read/write primitives */
 #define REG_WRITE(_ah, _reg, _val) \
--- a/drivers/net/wireless/ath/regd.c
+++ b/drivers/net/wireless/ath/regd.c
@@ -63,6 +63,9 @@ static struct reg_dmn_pair_mapping *ath_
 #define ATH_5GHZ_NO_MIDBAND	ATH_5GHZ_5150_5350, \
 				ATH_5GHZ_5725_5850
 
+/* new rule for ITS-G5 frequencies */
+#define ATH_5GHZ_ITS_G5         REG_RULE(5860-10, 5920+10, 10, 0, 33, NL80211_RRF_NO_IBSS)
+
 #define REGD_RULES(...) \
 	.reg_rules = { __VA_ARGS__ }, \
 	.n_reg_rules = ARRAY_SIZE(((struct ieee80211_reg_rule[]) { __VA_ARGS__ }))
