--- a/drivers/net/wireless/ath/ath9k/common-init.c
+++ b/drivers/net/wireless/ath/ath9k/common-init.c
@@ -86,6 +86,17 @@ static const struct ieee80211_channel at
 	CHAN5G(5785, 35), /* Channel 157 */
 	CHAN5G(5805, 36), /* Channel 161 */
 	CHAN5G(5825, 37), /* Channel 165 */
+	/* 802.11p patch */
+	/* ITS-G5B channels */
+	CHAN5G(5860, 38), /* Channel 172 - G5-SCH4 */
+	CHAN5G(5870, 39), /* Channel 174 - G5-SCH3 */
+	/* ITS-G5A channels */
+	CHAN5G(5880, 40), /* Channel 176 - G5-SCH1 */
+	CHAN5G(5890, 41), /* Channel 178 - G5-SCH2 */
+	CHAN5G(5900, 42), /* Channel 180 - G5-CCH  */
+	/* ITS-G5D channels */
+	CHAN5G(5910, 43), /* Channel 182 - G5-SCH5 */
+	CHAN5G(5920, 44), /* Channel 184 - G5-SCH6 */
 };
 
 /* Atheros hardware rate code addition for short preamble */
--- a/drivers/net/wireless/ath/ath9k/hw.h
+++ b/drivers/net/wireless/ath/ath9k/hw.h
@@ -74,7 +74,8 @@
 
 #define ATH9K_RSSI_BAD			-128
 
-#define ATH9K_NUM_CHANNELS	38
+/* 802.11p patch */ /* numbers of channels increased for ITS-G5 channels */
+#define ATH9K_NUM_CHANNELS	45
 
 /* Register read/write primitives */
 #define REG_WRITE(_ah, _reg, _val) \
--- a/drivers/net/wireless/ath/regd.c
+++ b/drivers/net/wireless/ath/regd.c
@@ -47,21 +47,24 @@ static struct reg_dmn_pair_mapping *ath_
 #define ATH_5GHZ_5150_5350	REG_RULE(5150-10, 5240+10, 80, 0, 30, 0),\
 				REG_RULE(5260-10, 5350+10, 80, 0, 30,\
 					 NL80211_RRF_NO_IR)
-#define ATH_5GHZ_5470_5850	REG_RULE(5470-10, 5850+10, 80, 0, 30,\
+/* 802.11p patch */ /* expanding 5 GHz frequences band up to 5.925 GHz */
+#define ATH_5GHZ_5470_5925	REG_RULE(5470-10, 5925+10, 80, 0, 30,\
 					 NL80211_RRF_NO_IR)
-#define ATH_5GHZ_5725_5850	REG_RULE(5725-10, 5850+10, 80, 0, 30,\
+#define ATH_5GHZ_5725_5925	REG_RULE(5725-10, 5925+10, 80, 0, 30,\
 					 NL80211_RRF_NO_IR)
 
 #define ATH_2GHZ_ALL		ATH_2GHZ_CH01_11, \
 				ATH_2GHZ_CH12_13, \
 				ATH_2GHZ_CH14
 
+/* 802.11p patch */ /* expanding 5 GHz frequences band up to 5.925 GHz */
 #define ATH_5GHZ_ALL		ATH_5GHZ_5150_5350, \
-				ATH_5GHZ_5470_5850
+				ATH_5GHZ_5470_5925
 
 /* This one skips what we call "mid band" */
+/* 802.11p patch */ /* expanding 5 GHz frequences band up to 5.925 GHz */
 #define ATH_5GHZ_NO_MIDBAND	ATH_5GHZ_5150_5350, \
-				ATH_5GHZ_5725_5850
+				ATH_5GHZ_5725_5925
 
 #define REGD_RULES(...) \
 	.reg_rules = { __VA_ARGS__ }, \
